File: backend/app/schemas/zabbix.py
"""
Mission Control Zabbix Schemas

Pydantic request/response models for Zabbix monitoring operations.

Sprint 2.3.0 - Enterprise Zabbix Integration.
"""

from pydantic import BaseModel

# ------------------------------------------------------------------ #
# Common                                                              #
# ------------------------------------------------------------------ #


class ZabbixConnectionTestResponse(BaseModel):
    """Response for connection test."""

    connected: bool
    latency_ms: int | None = None
    message: str | None = None
    version: str | None = None
    error: str | None = None


# ------------------------------------------------------------------ #
# Hosts                                                               #
# ------------------------------------------------------------------ #


class ZabbixHost(BaseModel):
    """Zabbix host."""

    hostid: str
    host: str
    name: str
    status: str
    available: str
    interface: str = ""
    groups: list[str] = []
    templates: list[str] = []
    last_access: str | None = None


class ZabbixHostsResponse(BaseModel):
    """Response for host listing."""

    connected: bool
    hosts: list[ZabbixHost] = []
    total_count: int = 0
    enabled_count: int = 0
    disabled_count: int = 0
    available_count: int = 0
    unavailable_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Host Groups                                                         #
# ------------------------------------------------------------------ #


class ZabbixHostGroup(BaseModel):
    """Zabbix host group."""

    groupid: str
    name: str
    host_count: int = 0


class ZabbixHostGroupsResponse(BaseModel):
    """Response for host group listing."""

    connected: bool
    groups: list[ZabbixHostGroup] = []
    total_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Triggers                                                            #
# ------------------------------------------------------------------ #


class ZabbixTrigger(BaseModel):
    """Zabbix trigger."""

    triggerid: str
    description: str
    status: str
    priority: str
    value: str
    hosts: list[str] = []


class ZabbixTriggersResponse(BaseModel):
    """Response for trigger listing."""

    connected: bool
    triggers: list[ZabbixTrigger] = []
    total_count: int = 0
    enabled_count: int = 0
    disabled_count: int = 0
    problem_count: int = 0
    ok_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Problems                                                            #
# ------------------------------------------------------------------ #


class ZabbixProblem(BaseModel):
    """Zabbix problem."""

    eventid: str
    name: str
    severity: str
    status: str
    acknowledged: bool
    host: str
    timestamp: str


class ZabbixProblemsResponse(BaseModel):
    """Response for problem listing."""

    connected: bool
    problems: list[ZabbixProblem] = []
    total_count: int = 0
    severity_counts: dict[str, int] = {}
    acknowledged_count: int = 0
    unacknowledged_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Events                                                              #
# ------------------------------------------------------------------ #


class ZabbixEvent(BaseModel):
    """Zabbix event."""

    eventid: str
    name: str
    severity: str
    status: str
    host: str
    timestamp: str


class ZabbixEventsResponse(BaseModel):
    """Response for event listing."""

    connected: bool
    events: list[ZabbixEvent] = []
    total_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Items                                                               #
# ------------------------------------------------------------------ #


class ZabbixItem(BaseModel):
    """Zabbix item."""

    itemid: str
    name: str
    key_: str = ""
    status: str
    type: str
    last_value: str = ""
    host: str = ""


class ZabbixItemsResponse(BaseModel):
    """Response for item listing."""

    connected: bool
    items: list[ZabbixItem] = []
    total_count: int = 0
    supported_count: int = 0
    unsupported_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Templates                                                           #
# ------------------------------------------------------------------ #


class ZabbixTemplate(BaseModel):
    """Zabbix template."""

    templateid: str
    name: str
    hosts_count: int = 0


class ZabbixTemplatesResponse(BaseModel):
    """Response for template listing."""

    connected: bool
    templates: list[ZabbixTemplate] = []
    total_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Dashboards                                                          #
# ------------------------------------------------------------------ #


class ZabbixDashboard(BaseModel):
    """Zabbix dashboard."""

    dashboardid: str
    name: str
    display_name: str = ""
    owner: str = ""
    pages: int = 0


class ZabbixDashboardsResponse(BaseModel):
    """Response for dashboard listing."""

    connected: bool
    dashboards: list[ZabbixDashboard] = []
    total_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Maps                                                                #
# ------------------------------------------------------------------ #


class ZabbixMap(BaseModel):
    """Zabbix map."""

    sysmapid: str
    name: str
    width: int = 0
    height: int = 0
    elements: int = 0


class ZabbixMapsResponse(BaseModel):
    """Response for map listing."""

    connected: bool
    maps: list[ZabbixMap] = []
    total_count: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Health                                                              #
# ------------------------------------------------------------------ #


class ZabbixDatabaseHealth(BaseModel):
    """Zabbix database health."""

    status: str = "unknown"
    type: str = "unknown"
    size_mb: int = 0


class ZabbixHealthResponse(BaseModel):
    """Response for Zabbix health."""

    connected: bool
    status: str = "unknown"
    version: str = "unknown"
    server: str = ""
    uptime_hours: int = 0
    api_latency_ms: int = 0
    database: ZabbixDatabaseHealth = ZabbixDatabaseHealth()
    proxy_count: int = 0
    poller_items_per_sec: int = 0
    trigger_functions_per_sec: int = 0
    error: str | None = None


# ------------------------------------------------------------------ #
# Summary                                                             #
# ------------------------------------------------------------------ #


class ZabbixSummaryResponse(BaseModel):
    """Response for Zabbix summary."""

    connected: bool
    version: str = "unknown"
    server_name: str = ""
    host_count: int = 0
    problem_count: int = 0
    critical_count: int = 0
    warning_count: int = 0
    ok_count: int = 0
    uptime_hours: int = 0
    api_latency_ms: int = 0
    error: str | None = None

File: frontend/src/services/zabbix.ts
import { apiClient } from "../utils/apiClient";

const API_BASE = "/api/v1/zabbix";

export interface ZabbixConnectionTestResult {
    connected: boolean;
    latency_ms: number | null;
    message: string | null;
    version: string | null;
    error: string | null;
}

export interface ZabbixSummary {
    connected: boolean;
    version: string;
    server_name: string;
    host_count: number;
    problem_count: number;
    critical_count: number;
    warning_count: number;
    ok_count: number;
    uptime_hours: number;
    api_latency_ms: number;
    error: string | null;
}

export interface ZabbixHost {
    hostid: string;
    host: string;
    name: string;
    status: string;
    available: string;
    interface: string;
    groups: string[];
    templates: string[];
    last_access: string | null;
}

export interface ZabbixHostsResponse {
    connected: boolean;
    hosts: ZabbixHost[];
    total_count: number;
    enabled_count: number;
    disabled_count: number;
    available_count: number;
    unavailable_count: number;
    error: string | null;
}

export interface ZabbixHostGroup {
    groupid: string;
    name: string;
    host_count: number;
}

export interface ZabbixHostGroupsResponse {
    connected: boolean;
    groups: ZabbixHostGroup[];
    total_count: number;
    error: string | null;
}

export interface ZabbixTrigger {
    triggerid: string;
    description: string;
    status: string;
    priority: string;
    value: string;
    hosts: string[];
}

export interface ZabbixTriggersResponse {
    connected: boolean;
    triggers: ZabbixTrigger[];
    total_count: number;
    enabled_count: number;
    disabled_count: number;
    problem_count: number;
    ok_count: number;
    error: string | null;
}

export interface ZabbixProblem {
    eventid: string;
    name: string;
    severity: string;
    status: string;
    acknowledged: boolean;
    host: string;
    timestamp: string;
}

export interface ZabbixProblemsResponse {
    connected: boolean;
    problems: ZabbixProblem[];
    total_count: number;
    severity_counts: Record<string, number>;
    acknowledged_count: number;
    unacknowledged_count: number;
    error: string | null;
}

export interface ZabbixEvent {
    eventid: string;
    name: string;
    severity: string;
    status: string;
    host: string;
    timestamp: string;
}

export interface ZabbixEventsResponse {
    connected: boolean;
    events: ZabbixEvent[];
    total_count: number;
    error: string | null;
}

export interface ZabbixItem {
    itemid: string;
    name: string;
    key_: string;
    status: string;
    type: string;
    last_value: string;
    host: string;
}

export interface ZabbixItemsResponse {
    connected: boolean;
    items: ZabbixItem[];
    total_count: number;
    supported_count: number;
    unsupported_count: number;
    error: string | null;
}

export interface ZabbixTemplate {
    templateid: string;
    name: string;
    hosts_count: number;
}

export interface ZabbixTemplatesResponse {
    connected: boolean;
    templates: ZabbixTemplate[];
    total_count: number;
    error: string | null;
}

export interface ZabbixDashboard {
    dashboardid: string;
    name: string;
    display_name: string;
    owner: string;
    pages: number;
}

export interface ZabbixDashboardsResponse {
    connected: boolean;
    dashboards: ZabbixDashboard[];
    total_count: number;
    error: string | null;
}

export interface ZabbixMap {
    sysmapid: string;
    name: string;
    width: number;
    height: number;
    elements: number;
}

export interface ZabbixMapsResponse {
    connected: boolean;
    maps: ZabbixMap[];
    total_count: number;
    error: string | null;
}

export interface ZabbixDatabaseHealth {
    status: string;
    type: string;
    size_mb: number;
}

export interface ZabbixHealthResponse {
    connected: boolean;
    status: string;
    version: string;
    server: string;
    uptime_hours: number;
    api_latency_ms: number;
    database: ZabbixDatabaseHealth;
    proxy_count: number;
    poller_items_per_sec: number;
    trigger_functions_per_sec: number;
    error: string | null;
}

export const zabbixApi = {
    testConnection: () => apiClient<ZabbixConnectionTestResult>(`${API_BASE}/test`, { method: "POST" }),
    getOverview: () => apiClient<ZabbixSummary>(`${API_BASE}/overview`),
    getHosts: () => apiClient<ZabbixHostsResponse>(`${API_BASE}/hosts`),
    getGroups: () => apiClient<ZabbixHostGroupsResponse>(`${API_BASE}/groups`),
    getTemplates: () => apiClient<ZabbixTemplatesResponse>(`${API_BASE}/templates`),
    getItems: () => apiClient<ZabbixItemsResponse>(`${API_BASE}/items`),
    getTriggers: () => apiClient<ZabbixTriggersResponse>(`${API_BASE}/triggers`),
    getProblems: () => apiClient<ZabbixProblemsResponse>(`${API_BASE}/problems`),
    getEvents: () => apiClient<ZabbixEventsResponse>(`${API_BASE}/events`),
    getMaps: () => apiClient<ZabbixMapsResponse>(`${API_BASE}/maps`),
    getDashboards: () => apiClient<ZabbixDashboardsResponse>(`${API_BASE}/dashboards`),
    getHealth: () => apiClient<ZabbixHealthResponse>(`${API_BASE}/health`),
};
